### PR TITLE
Detect type equiv for System.Tuple`8

### PR DESCRIPTION
**Pull request by gusty**
This will allow the constraint solver to detect type equivalences between compiled and syntactic tuples over 8 elements.

### CODE REVIEW

**Comment by NinoFloris** [comment]
Welcome fix! Running into this as well @cartermp

**Comment by cartermp** [comment]
Looks like this addresses https://github.com/dotnet/fsharp/issues/9639 ? @gusty do you still intend on this being a draft?

File: tests/fsharp/core/csext/test.fsx [hunk 1]
```diff
@@ -311,6 +311,9 @@ module TupleSRTP =
     let v1b =  (^T : (member get_Item2 : unit -> _ ) (new System.Tuple<int,int>(1,3)))
     let v2b =  (^T : (member get_Item2 : unit -> _ ) (System.Tuple<int,int>(1,3)))
     let v3b =  (^T : (member get_Item2 : unit -> _ ) ((1,3)))
+    
+    let areEqualT8 = System.Tuple<_,_,_,_,_,_,_,_>(1,2,3,4,5,6,7,System.Tuple<_>  (8)  ) = (1,2,3,4,5,6,7,8)
```
Comment (NinoFloris) [inline tests/fsharp/core/csext/test.fsx:315]: Will this also compile for:

```
     let areEqualT9 = System.Tuple<_,_,_,_,_,_,_,(int * int)>(1,2,3,4,5,6,7,(8,9)) = (1,2,3,4,5,6,7,8,9) 
```

**Comment by gusty** [comment]
@cartermp I did this as part of #9605 but realized that it’s really a fix of an inconsistent behavior of the compiler, so I wouldn’t be surprised that someone else encountered this issue.

It’s in draft because I’m not 100% sure it’s complete. There might be more cases I’m missing.

I’ll have another look later and let you know.

File: tests/fsharp/core/csext/test.fsx (hunk 1 above)
Comment (gusty) [reply tests/fsharp/core/csext/test.fsx:315]: Yes, I added the 8-uple case as it was trickier to get it right, but it should work for any tuple size.

Comment (gusty) [reply tests/fsharp/core/csext/test.fsx:315]: I just added a commit that will allow that.

**Comment by NinoFloris** [comment]
About https://github.com/dotnet/fsharp/pull/9586/commits/c61e3ad7f0b0eafa10c6a3e436fcda598aafd8cc, unless I read it wrong I think we should strongly normalize System.Tuple to syntactic tuples but not vice versa. Allowing tuples at 8th position in syntactic tuples to stand in for `TRest` does not sound right. In such cases `TRest` should be `Tuple<(int * int)>` not `Tuple<int, int>`.

**Comment by gusty** [comment]
I wish we could convert all inferred tuples to syntactic tuples but that's not possible or it's too complicated.

So, our best alternative is enforcing type equivalence, which is what this PR does, now syntactic tuples and system tuples could appear at any side of the equality.

Your code, won't compile if we don't have the other way around.

Again, if we're talking about normalizing I agree with you, but that's not what this PR does.

**Comment by TIHan** [comment]
When we type-check an application, `System.Tuple` gets decompiled into TType_tuple.
```fsharp
    // Try to decode System.Tuple --> F~ tuple types etc.
    let ty = g.decompileType tcref actualArgTys
```
Are we not able to do the same here? This is how we started trying to normalize tuples.

**Comment by dsyme** [comment]
> If you're saying we shouldn't implement any fix here then I see no generic way to deal with syntactic tuples, wouldn't opting into using the encoding be enough of a signal you're coupling yourself to .NET?

That's right, in the current design of F# there is no intention to give a statically-resolved way to write generic inlined SRTP code that works over any tuple type.

If we did support this I think it would be an extension to SRTP constraints with respect to the F# source tuple types, not the encoding.

**Comment by NinoFloris** [comment]
I expected as much, and I'm 'ok' with what it's doing right now.

For the curious readers thinking "why would you ever write or need this"; this code is used for an inference driven reader CE. A custom operation on the builder enables you to request common dependencies outside function arguments. It's really quite a joy to use.

```fs
// inferred type ContextTask<'a, Widget voption> when 'a :> DbConnection and 'a :> Logging and 'a :> Tenant
let findWidget widgetId = contextTask {
    arguments (Service.dbConnection, Service.logging, Service.tenant) into (dbConn, logger, tenant)
    try 
        return! dbConn.TrySingle<Widget>("SELECT * FROM widgets WHERE id = @widgetId AND tenantId = @tenantId", {| widgetId = widgetId; tenantId = tenant.Id |})
    with ex -> 
        return logger.LogError("Oh no", ex)
} 
```

To do that `arguments` must be polyvariadic (infinite differently typed args) to allow for the separate return types of the dependencies, and as a result `TupleApp` was born.
For completeness sake I had hoped to have it function identically for all argument lengths but I can live with the 7 arguments 'until it gets unpleasant' limit, more should be viewed as a code smell anyway.

**Comment by dsyme** [comment]
> I expected as much, and I'm 'ok' with what it's doing right now.

OK cool.  I will close this then, thank you for the example too

**Comment by NinoFloris** [comment]
>If we did support this I think it would ideally be an extension to SRTP constraints with respect to the F# source tuple types, not the encoding

Hypothetically what would this look like? I expect some primitive type level pattern matching would be necessary. 
However how to inductively construct (and with minimal additional syntax express) a flat n-tuple is something I'm still not clear on. This gap in particular is the crux of why we're abusing the underlying encoding.

**Comment by gusty** [comment]
I think we all agree that the sole action of declaring this behavior as by-design doesn't automatically make it more consistent.

I was thinking, in order to solve this inconsistency at some point sooner or later we'll need to add a new static constraint that indicates that a type is a tuple of a yet unknown arity.

So if we take @dsyme example, typed `inline`, it would give:

```
let inline f x = System.Tuple<_,_,_,_,_,_,_,_>(1,2,3,4,5,6,7,x) 
// val inline f : x:'a -> System.Tuple<int,int,int,int,int,int,int,'a when 'a : tuple>
```

then there's no possibility of creating an invalid tuple, and at the same time we can, at the very last step of type inference, convert all tuples of known arity to a syntactic one. This way, when calling the above function, we'll get:

```
> f (System.Tuple<_> (8)) ;;
val it : int * int* int* int* int* int* int* int =
  (1, 2, 3, 4, 5, 6, 7, 8)
```

instead of `val it : System.Tuple<int,int,int,int,int,int,int,System.Tuple<int>>`

So, supplying a generic parameter won't resolve the constraint and convert to syntactic, but if we supply a tuple, even if it's generic but its arity is know, like `'a * 'b` it resolves, unless of course it's another 8-non-encoded tuple.

Another constraint should be added for struct tuples: `^T when ^T : struct tuple`.

**Pull request closed without merging**